src/org/traccar/protocol/TeltonikaProtocolDecoder.java

```diff
@@ -240,27 +244,57 @@ private void decodeLocation(Position position, ChannelBuffer buf, int codec) {
         }
 
         if (channel != null) {
-            ChannelBuffer response = ChannelBuffers.directBuffer(4);
-            response.writeInt(count);
-            channel.write(response);
+            if (connectionless) {
+                ChannelBuffer response = ChannelBuffers.directBuffer(5);
+                response.writeShort(3);
+                response.writeShort(packetId);
+                response.writeByte(0x02);
+                channel.write(response, remoteAddress);
+            } else {
+                ChannelBuffer response = ChannelBuffers.directBuffer(4);
+                response.writeInt(count);
+                channel.write(response);
+            }
         }
 
         return positions;
     }
 
     @Override
-    protected Object decode(
-            Channel channel, SocketAddress remoteAddress, Object msg) throws Exception {
+    protected Object decode(Channel channel, SocketAddress remoteAddress, Object msg) throws Exception {
 
         ChannelBuffer buf = (ChannelBuffer) msg;
 
+        if (channel instanceof DatagramChannel) {
+            return decodeUDP(channel, remoteAddress, buf);
+        } else {
+            return decodeTCP(channel, remoteAddress, buf);
+        }
+    }
+
+    protected Object decodeTCP(Channel channel, SocketAddress remoteAddress, ChannelBuffer buf) throws Exception {
+
         if (buf.getUnsignedShort(0) > 0) {
             parseIdentification(channel, remoteAddress, buf);
         } else {
-            return parseData(channel, remoteAddress, buf);
+            return parseData(channel, remoteAddress, buf, false, 4, (short) 0);
         }
 
         return null;
     }
 
+    protected Object decodeUDP(Channel channel, SocketAddress remoteAddress, ChannelBuffer buf) throws Exception {
+        int udpPacketLength = buf.getUnsignedShort(0);
```

Comment:
I think you should simplify variable names. For example, it doesn't make sense to have a `udp` prefix because the method is for UDP only.